### PR TITLE
Adding ethereum witness contracts

### PR DESCRIPTION
**Pull request by sinasabet81**
This PR adds the contracts for the Ethereum Witness. The files of interest are the Solidity contracts, under `contracts/`, and their Python tests under `tests/`.

See the README for info on how to run the tests locally.

### CODE REVIEW

File: witness/ethereum/README.md
Comment (smeiklej) [inline witness/ethereum/README.md:44]: "Explanation"

File: witness/ethereum/README.md
Comment (smeiklej) [inline witness/ethereum/README.md:46]: Should be Merkle (not merkle).

File: witness/ethereum/README.md
Comment (smeiklej) [inline witness/ethereum/README.md:51]: Same comment, should be Merkle.

Comment (smeiklej) [inline witness/ethereum/README.md:51]: Small thought: I think the higher-level readme we're planning is the right place to elaborate on the differences between using "normal" consistency proofs and using compact ranges, but it might be worth adding a quick note here saying that `RootWitness` and `CompactRangeWitness` are separate contracts that verify consistency in two different ways?  We can also wait to add this until we have the higher-level readme in place if that makes more sense.

**Review by smeiklej — APPROVED** [review]
LGTM modulo a few small comments about the readme.

File: witness/ethereum/README.md
Comment (sinasabet81) [reply witness/ethereum/README.md]: I updated this section; please take a look and let me know what you think!

**Comment by sinasabet81** [comment]
@smeiklej just pushed an update with the fixes from your review; please take a look!

File: witness/ethereum/README.md
Comment (smeiklej) [reply witness/ethereum/README.md]: Looks good to me!

**Review by pavelkalinnikov — APPROVED** [review]

**Comment by mhutchinson** [comment]
This PR has added a number of dependencies which are showing as having security vulns. Can someone take a look at https://github.com/google/trillian-examples/security/dependabot as a matter of urgency?

**Comment by sinasabet81** [comment]
Sorry about that, looking into upgrading those versions in another PR now.

**Comment by sinasabet81** [comment]
PR up bumping the package versions @ https://github.com/google/trillian-examples/pull/376

**Comment by sinasabet81** [comment]
#367 seems to have fixed some of the issues, but other dependencies are still out of date; I think we're stuck here for now, since the eth-brownie repo seems to specify the minor version of these deps (eg. `urllib` in @ https://github.com/eth-brownie/brownie/blob/master/requirements.txt).

I'll follow up with the eth-brownie peeps and see if they can bump their dependencies, but even then we'll likely have to wait for them to cut a fresh release.

File: witness/ethereum/Dockerfile
Comment (ernestognw) [inline witness/ethereum/Dockerfile:13]: Adding global might have the same issues as with the `pip install`. I'm pretty sure you can lock to an specific ganache version by adding a `package.json`, the auto generated `yarn.lock` and running locally `yarn add ganache-cli`

Eventually if it's needed to run, it can be achieved with `npx ganache-cli` or `yarn run ganache-cli`

File: witness/ethereum/contracts/RootWitness.sol
Comment (ernestognw) [inline witness/ethereum/contracts/RootWitness.sol:98]: Are the timestamps possibly optimizable in terms of size?

A `uint32` gives enough room for Unix epoch timestamps in case this is the format desired. Otherwise `uint40` is more than enough.

File: witness/ethereum/contracts/CompactRangeWitness.sol
Comment (ernestognw) [inline witness/ethereum/contracts/CompactRangeWitness.sol:48]: I guess this `Ownable` logic can be moved to another contract, it's repeated in both witness

File: witness/ethereum/contracts/CompactRangeWitness.sol
Comment (ernestognw) [inline witness/ethereum/contracts/CompactRangeWitness.sol:81]: Maybe change to a `modifier`

**Review by ernestognw — COMMENTED** [review]
Hi @sinasabet81

I just left some small nits to my understanding.

Great to see something like this

**Pull request merged**
